sdk/azidentity/chained_token_credential.go

```
@@ -17,43 +18,46 @@ type ChainedTokenCredential struct {
 	var errList []*CredentialUnavailableError
-	for i := 0; i < len(c.sources); i++ {
+	for _, cred := range c.sources { // loop through all of the credentials provided in sources
+		token, err = cred.GetToken(ctx, opts) // make a GetToken request for the current credential in the loop
 		var credErr *CredentialUnavailableError
-		token, err = c.sources[i].GetToken(ctx, opts)
-		if errors.As(err, &credErr) {
-			errList = append(errList, credErr)
-		} else if err != nil {
-			errList = append(errList, &CredentialUnavailableError{CredentialType: "Chained Token Credential", Message: err.Error()})
-			break
+		if errors.As(err, &credErr) { // check if we received a CredentialUnavailableError
+			errList = append(errList, credErr) // if we did receive a CredentialUnavailableError then we append it to our error slice and continue looping for a good credential
+		} else if err != nil { // if we receive some other type of error then we must stop looping and process the error accordingly
+			var authenticationFailed *AuthenticationFailedError
+			if errors.As(err, &authenticationFailed) { // if the error is an AuthenticationFailedError we return the error related to the invalid credential and append all of the other error messages received prior to this point
+				return nil, &AuthenticationFailedError{msg: "Received an AuthenticationFailedError, there is an invalid credential in the chain. " + createChainedErrorMessage(errList), inner: err}
+			}
+			return nil, fmt.Errorf("Received an unexpected error: %w", err) // if we receive some other error type this is unexpected and we simple return the unexpected error
 		} else {
-			return token, nil
+			return token, nil // if we did not receive an error then we return the token
 		}
 	}
-	// This condition should never be true
+	// this situation can occur when the length of sources is 0
 	if token == nil && len(errList) == 0 {
 		return nil, nil
```

Perhaps this should return an error that no token was returned rather than simply also returning nil? If we think that's the case you could modify `createChainedErrorMessage` to return a special message when errList is empty, then you could just remove this condition entirely.

---

Agreed an error should be returned here.